### PR TITLE
New flag to limit plugins to single instances

### PR DESCRIPTION
**Pull request by paulolieuthier**
@palinek which do you think it's the cleanest way to get the flag out of the plugin in the AddPluginDialog class?

Fixes lxde/lxqt#637.

### CODE REVIEW

**Comment by palinek** [comment]
@paulolieuthier when you assigned me... do you want me to implement the logic or just the answer for the question above?

**Comment by paulolieuthier** [comment]
TBH, I don't really remember.

**Comment by paulolieuthier** [comment]
@palinek if you can, please review it and make your always helpful comments.

**Comment by paulolieuthier** [comment]
@palinek done. Maybe there is a cleaner way to do it, though.

**Comment by paulolieuthier** [comment]
@palinek you can review now. :smile:

File: panel/panelpluginsmodel.cpp [hunk 1]
```diff
@@ -107,11 +107,23 @@ QList<Plugin *> PanelPluginsModel::plugins() const
...
 void PanelPluginsModel::addPlugin(const LxQt::PluginInfo &desktopFile)
 {
     QString name = findNewPluginSettingsGroup(desktopFile.id());
     QPointer<Plugin> plugin = loadPlugin(desktopFile, name);
-    if (!plugin.isNull())
+    if (plugin.isNull())
+        return;
+
+    bool single = plugin->iPlugin()->flags().testFlag(ILxQtPanelPlugin::SingleInstance);
+    if (!single || !pluginNames().contains(desktopFile.id()))
```
Comment (palinek) [inline panel/panelpluginsmodel.cpp]: Does this even work for you? Here the 'type' is searched plugins names (instances) => can't be found there.

Actualy while reviewing this I think that the `SingleInstance` flag should be known before constucting the Plugin object => somewhere in `ILxQtPanelPlugin` to not create next object if one already created.

Comment (paulolieuthier) [inline panel/panelpluginsmodel.cpp]: > Does this even work for you? Here the 'type' is searched plugins names (instances) => can't be found there.

I don't follow you here. But it does work for me.

> Actualy while reviewing this I think that the SingleInstance flag should be known before constructing the Plugin object => somewhere in ILxQtPanelPluginLibrary to not create next object if one already created.

That would be great, but I don't know how to do it.

Comment (paulolieuthier) [inline panel/panelpluginsmodel.cpp]: Please, feel free to modify it as you think it's best and merge.

**Comment by jleclanche** [comment]
So which plugins would be SingleInstance?

**Comment by paulolieuthier** [comment]
@jleclanche for now, the system tray plugin.

**Comment by jleclanche** [comment]
Just the x11 one right? Because there's no reason it should be single instance, other than the xembed limitations.

Which makes me wonder why we need a flag; Imo this is probably best to hardcode in the xembed plugin.

File: panel/panelpluginsmodel.cpp (hunk 1 above)
Comment (palinek) [inline panel/panelpluginsmodel.cpp]: >> Does this even work for you? Here the 'type' is searched plugins names (instances) => can't be found there.

>I don't follow you here. But it does work for me.

`pluginNames()` are instances -> e.g: 'mainmenu2', 'spacer2', 'spacer3', 'tray2' etc.. in contrast with `desktopFile.id()` which is base name of the .desktop file -> e.g.: 'mainmenu', 'spacer', 'tray'

You are just lucky that your 'tray' plugin instance doesn't have autogenerated name. Try to remove the plugin from panel and reinsert it after that.


>> Actualy while reviewing this I think that the SingleInstance flag should be known before constructing the Plugin object => somewhere in ILxQtPanelPluginLibrary to not create next object if one already created.

> That would be great, but I don't know how to do it.

It must be known somewhere here https://github.com/lxde/lxqt-panel/blob/master/panel/plugin.cpp#L251 => so maybe implemented directly in `instance()` method ... that means specific plugin implementation as also @jleclanche suggested.

**Comment by paulolieuthier** [comment]
Actually, the status notifier plugin [as well](http://www.notmart.org/misc/statusnotifieritem/statusnotifierwatcher.html). We shouldn't hardcode it, because the user wouldn't know why he can't add a duplicate system tray or status notifier.

**Comment by paulolieuthier** [comment]
@palinek: updated and rebased. I hope it's better now.

**Comment by paulolieuthier** [comment]
@palinek about the issue with multiple panels, we can get information about all panels easily using something like:

```c++
LxQtPanelApplication *app = reinterpret_cast<LxQtPanelApplication*>(qApp);
for (LxQtPanel *panel : app->mPanels)
{
    panel->mPlugins ...
}
```

But it looks really ugly. How do you think it'd be better to approach that issue?

**Comment by palinek** [comment]
> But it looks really ugly. How do you think it'd be better to approach that issue?

We can provide something like `LxQtPanelApplication::isSingletonAndRunningPlugin(QString const & pluginId)` which will simple test all panels internaly...

**Comment by palinek** [comment]
@paulolieuthier working on the proposed solution...

**Comment by paulolieuthier** [comment]
Nice. :+1:

**Comment by paulolieuthier** [comment]
@palinek great work.

**Pull request merged**